### PR TITLE
[feature/emm] EMM version support

### PR DESCRIPTION
**Pull request by felix-schwarz**
## Description
This pull requests adds what's needed for a separate EMM version of the app:
- [x] globally available EMM detection via `OCLicenseEMMProvider.isEMMVersion` (detects an EMM version if the main bundle ID (as found in the app's or app extension's `Info.plist`) ends with either `-emm` or `.emm`)
- [x] new `OCLicenseEMMProvider` unlocking all pro features, if an EMM version is detected
- [x] removal of *Pro Features* and *Purchases* from Settings if an EMM version is detected
- [x] new fastlane lane for building the EMM version, using bundle identifiers for all the app and app extensions ending in either `-emm` or `.emm`
- [x] different app icon (with lock) to distinguish it in the App Store

### Fastlane

A new lane `release_on_appstore` was added. This lane is responsible for:
- create release notes from in-app release notes file `ownCloud/Release Notes/ReleaseNotes.plist` into the fastlane metadata file `fastlane/metadata/en-us/metadata.txt`
- create screenshots
- push changed files (release notes, screenshots) to git
- set bundle ids, provisioning profiles, build regular iOS app
- deliver build, screenshots and metadata to the App Store 
- set bundle ids, provisioning profiles, build EMM iOS app
- deliver build, screenshots and metadata to the App Store 

All this steps can be disabled, by setting parameters to the lane execution:

`create_release_notes:false` do not create release notes
`create_screenshots:false` do not create app screenshots
`build_regular_app:false` do not build the regular iOS app
`build_emm_app:false` do not build the EMM iOS app
`deliver:false` do not deliver the EMM iOS app to the App Store

**Sample:**
```
fastlane release_on_appstore create_release_notes:false create_screenshots:false build_regular_app:false build_emm_app:false deliver:false 
```

## Related Issue
https://github.com/owncloud/enterprise/issues/2961

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by felix-schwarz** [comment]
@hosy Idea: for support, it might be useful to have `-emm` (or something along those lines) added to the version info at the bottom of the Settings.

File: ownCloud/Tools/VendorServices.swift [hunk 1]
```diff
@@ -54,6 +54,14 @@ class VendorServices : NSObject {
 		return false
 	}
 
+	var isEMMVersion: Bool {
+		if let isEMMVersion = Bundle.main.bundleIdentifier?.hasSuffix(".emm") {
```
Comment (felix-schwarz) [inline ownCloud/Tools/VendorServices.swift:58]: For consistency, this should just return the value of `OCLicenseEMMProvider.isEMMVersion`, so there's only one check to adjust if it ever needs to be changed.

Comment (hosy) [reply ownCloud/Tools/VendorServices.swift]: Thanks, fixed!

**Comment by jesmrec** [comment]
Execution of 

` fastlane release_on_appstore create_release_notes:false create_screenshots:false build_regular_app:false build_emm_app:true deliver:false` after adding `.emm` to the bundle identifier:

```
+-----------------------+---------+-----------+
|                Used plugins                 |
+-----------------------+---------+-----------+
| Plugin                | Version | Action    |
+-----------------------+---------+-----------+
| fastlane-plugin-badge | 1.2.0   | add_badge |
+-----------------------+---------+-----------+

[14:19:46]: In the config file './fastlane/Matchfile' you have the line app_identifier, but didn't provide any value. Make sure to append a value right after the option name. Make sure to check the docs for more information
Updating plugin dependencies...
[DEPRECATED] `Bundler.with_clean_env` has been deprecated in favor of `Bundler.with_unbundled_env`. If you instead want the environment before bundler was originally loaded, use `Bundler.with_original_env` (called at /Users/jesus/.rvm/gems/ruby-2.4.1/gems/fastlane-2.135.1/fastlane/lib/fastlane/plugins/plugin_manager.rb:218)
Successfully updated plugins
➜  ios-app git:(feature/emm) ✗ fastlane release_on_appstore create_release_notes:false create_screenshots:false build_regular_app:false build_emm_app:true deliver:false
[✔] 🚀 
+-----------------------+---------+-----------+
|                Used plugins                 |
+-----------------------+---------+-----------+
| Plugin                | Version | Action    |
+-----------------------+---------+-----------+
| fastlane-plugin-badge | 1.4.0   | add_badge |
+-----------------------+---------+-----------+

[14:20:16]: In the config file './fastlane/Matchfile' you have the line app_identifier, but didn't provide any value. Make sure to append a value right after the option name. Make sure to check the docs for more information
[14:20:16]: ------------------------------
[14:20:16]: --- Step: default_platform ---
[14:20:16]: ------------------------------
[14:20:16]: Driving the lane 'ios release_on_appstore' 🚀
[14:20:16]: In the config file './Matchfile' you have the line app_identifier, but didn't provide any value. Make sure to append a value right after the option name. Make sure to check the docs for more information
[14:20:16]: Build EMM iOS App…
[14:20:16]: In the config file './Matchfile' you have the line app_identifier, but didn't provide any value. Make sure to append a value right after the option name. Make sure to check the docs for more information
[14:20:16]: --------------------------------------------------------------------------------------------------------------------------------------------
[14:20:16]: --- Step: mv ../ownCloud/Resources/Assets.xcassets/AppIcon.appiconset/ ../ownCloud/Resources/Assets.xcassets/AppIcon-regular.appiconset/ ---
[14:20:16]: --------------------------------------------------------------------------------------------------------------------------------------------
[14:20:16]: $ mv ../ownCloud/Resources/Assets.xcassets/AppIcon.appiconset/ ../ownCloud/Resources/Assets.xcassets/AppIcon-regular.appiconset/
[14:20:16]: ----------------------------------------------------------------------------------------------------------------------------------------
[14:20:16]: --- Step: mv ../ownCloud/Resources/Assets.xcassets/AppIcon-emm.appiconset/ ../ownCloud/Resources/Assets.xcassets/AppIcon.appiconset/ ---
[14:20:16]: ----------------------------------------------------------------------------------------------------------------------------------------
[14:20:16]: $ mv ../ownCloud/Resources/Assets.xcassets/AppIcon-emm.appiconset/ ../ownCloud/Resources/Assets.xcassets/AppIcon.appiconset/
[14:20:16]: ▸ mv: rename ../ownCloud/Resources/Assets.xcassets/AppIcon-emm.appiconset/ to ../ownCloud/Resources/Assets.xcassets/AppIcon.appiconset/: No such file or directory
+------------------+-------------------------+
|                Lane Context                |
+------------------+-------------------------+
| DEFAULT_PLATFORM | ios                     |
| PLATFORM_NAME    | ios                     |
| LANE_NAME        | ios release_on_appstore |
+------------------+-------------------------+
[14:20:16]: Exit status of command 'mv ../ownCloud/Resources/Assets.xcassets/AppIcon-emm.appiconset/ ../ownCloud/Resources/Assets.xcassets/AppIcon.appiconset/' was 1 instead of 0.
mv: rename ../ownCloud/Resources/Assets.xcassets/AppIcon-emm.appiconset/ to ../ownCloud/Resources/Assets.xcassets/AppIcon.appiconset/: No such file or directory


+------+---------------------------------------------------------------+-------------+
|                                  fastlane summary                                  |
+------+---------------------------------------------------------------+-------------+
| Step | Action                                                        | Time (in s) |
+------+---------------------------------------------------------------+-------------+
| 1    | default_platform                                              | 0           |
| 2    | mv ../ownCloud/Resources/Assets.xcassets/AppIcon.appiconset/  | 0           |
| 💥   | mv ../ownCloud/Resources/Assets.xcassets/AppIcon-emm.appicons | 0           |
+------+---------------------------------------------------------------+-------------+

[14:20:16]: fastlane finished with errors

Looking for related GitHub issues on fastlane/fastlane...

/Users/jesus/.rvm/gems/ruby-2.4.1/gems/fastlane-2.149.1/fastlane_core/lib/fastlane_core/ui/interface.rb:153:in `shell_error!': [!] Exit status of command 'mv ../ownCloud/Resources/Assets.xcassets/AppIcon-emm.appiconset/ ../ownCloud/Resources/Assets.xcassets/AppIcon.appiconset/' was 1 instead of 0. (FastlaneCore::Interface::FastlaneShellError)
mv: rename ../ownCloud/Resources/Assets.xcassets/AppIcon-emm.appiconset/ to ../ownCloud/Resources/Assets.xcassets/AppIcon.appiconset/: No such file or directory

```

`AppIcon-emm.appiconset` is not present. Am i missing any step?

**Comment by hosy** [comment]
> Execution of
> 
> ` fastlane release_on_appstore create_release_notes:false create_screenshots:false build_regular_app:false build_emm_app:true deliver:false` after adding `.emm` to the bundle identifier:
> 

@jesmrec do not add the `.emm` to the bundle identifier manually. This will be done by the fastlane script.
Please pull this branch from git and do not modify anything and run the command above.

If you run in the same problem, please post your fastlane version number.

**Comment by jesmrec** [comment]
I'm running `fastlane 2.149.1` 

Problem above is fixed with your instructions. I had also to upgrade plugins with `bundle exec fastlane update_plugins` before execution.

The fastlane command execution failed now with
```
[18:46:52]: Error setting value 'ownCloud.xcworkspace' for option 'workspace'
[18:46:52]: You passed invalid parameters to 'gym'.
[18:46:52]: Check out the error below and available options by running `fastlane action gym`
```
this is the whole output:

[fastlane_output.zip](https://github.com/owncloud/ios-app/files/4787941/fastlane_output.zip)

in which, the Workspace file `ownCloud.xcworkspace` is not found (it does not exist in my local).

**Comment by hosy** [comment]
@jesmrec this error is independent of the `EMM` version. Please run `pod install` before you run the `fastlane` command. This should create the missing file `ownCloud.xcworkspace `

**Comment by jesmrec** [comment]
By executing `pod install` before, the fastlane command finishes succesfully.

I tested two different scenarios:

- 1.- Install EMM version in a real device. Even when XCode builds correctly, the app can not be installed in the devices with the following error:

```
Details

Unable to install "ownCloud"
Domain: com.apple.dt.MobileDeviceErrorDomain
Code: -402620394
--
The executable was signed with invalid entitlements.
Domain: com.apple.dt.MobileDeviceErrorDomain
Code: -402620394
User Info: {
    DVTRadarComponentKey = 487925;
    MobileDeviceErrorCode = "(0xE8008016)";
    "com.apple.dtdevicekit.stacktrace" = (
	0   DTDeviceKitBase                     0x000000011c06681a DTDKCreateNSErrorFromAMDErrorCode + 233
	1   DTDeviceKitBase                     0x000000011c0a7f70 __90-[DTDKMobileDeviceToken installApplicationBundleAtPath:withOptions:andError:withCallback:]_block_invoke + 155
	2   DVTFoundation                       0x0000000102f0d385 DVTInvokeWithStrongOwnership + 73
	3   DTDeviceKitBase                     0x000000011c0a7ca8 -[DTDKMobileDeviceToken installApplicationBundleAtPath:withOptions:andError:withCallback:] + 1654
	4   IDEiOSSupportCore                   0x000000011bf1ee91 __118-[DVTiOSDevice(DVTiPhoneApplicationInstallation) processAppInstallSet:appUninstallSet:installOptions:completionBlock:]_block_invoke.352 + 4165
	5   DVTFoundation                       0x0000000103040a10 __DVT_CALLING_CLIENT_BLOCK__ + 7
	6   DVTFoundation                       0x0000000103042652 __DVTDispatchAsync_block_invoke + 1194
	7   libdispatch.dylib                   0x00007fff72a6b6c4 _dispatch_call_block_and_release + 12
	8   libdispatch.dylib                   0x00007fff72a6c658 _dispatch_client_callout + 8
	9   libdispatch.dylib                   0x00007fff72a71c44 _dispatch_lane_serial_drain + 597
	10  libdispatch.dylib                   0x00007fff72a725d6 _dispatch_lane_invoke + 363
	11  libdispatch.dylib                   0x00007fff72a7bc09 _dispatch_workloop_worker_thread + 596
	12  libsystem_pthread.dylib             0x00007fff72cc6a3d _pthread_wqthread + 290
	13  libsystem_pthread.dylib             0x00007fff72cc5b77 start_wqthread + 15
);
}
--


System Information

macOS Version 10.15.4 (Build 19E266)
Xcode 11.4 (16134)
```

Devices are: iPhoneXR and iPadAir2, both of them with iOS13.4

- 2.- Install EMM version in an emulator. XCode builds correctly the app, and it is correctly installed (iPhone8):

Screenshot 2020-06-23 at 09 36 10

At this point, the Pro Features are unlocked and the `Settings` does not show such section. This is also correct.

About the Pro features:

- Document scanner: Not posible to test in an emulator, because the lack of camera. App crashes when selecting Document Scanner (anything to improve like hidding the option if no camera is detected? if feasible, to do in a separate issue, not here) 

- Shortcuts: Not working. Error after executing any shortcut. For example, creation of a folder:

Jun-23-2020 09-38-04

- Markup: File is not displayed, no matter if picture or PDF:

|Image|PDF|
|:-:|:-:|
|Jun-23-2020 09-41-14|Jun-23-2020 09-54-29|

I have tested the same scenarios in the regular app (`master` version), with the same account and the same files. it worked correctly.

**Comment by hosy** [comment]
@jesmrec in my opinion all your findings are not a problem of this branch and should be tested in `milestone/1.4` again.
regarding installing the version on a device, this may be a problem of installing an App Store signed version.

**Comment by jesmrec** [comment]
> @jesmrec in my opinion all your findings are not a problem of this branch and should be tested in milestone/1.4 again.

In `master` it's working fine
In `milestone/1.4 (385f592)` it's working fine

i will check again as part of the 11.4 regression, for sure. But, i do not have the certainty that this is working correctly in this EMM version. Afaiu, the EMM version will be available in App Store. If this is available in TestFlight mode, i can test there again with a real device.

**Comment by hosy** [comment]
@jesmrec yes, but there are no code changes regarding all this problems. `EMM` are only removing the pay wall for this features.
Sometimes I also noticed this problems, but not in this branch and it could also depend on connectivity or core problems.

**Comment by hosy** [comment]
@jesmrec yes, I think it make sense to test this all in the TestFlight build.

**Comment by jesmrec** [comment]
For me it's ok to move forward, since you required me just to test that IAPs are disabled (features enabled), and the label on the bottom of the Settings view is correct with EMM suffix.

Please, ping me when EMM version is available in TestFlight, to check the pro features there.

**Comment by jesmrec** [comment]
Ok, let's move this forward then, keeping an eye on the Testflight when it is.

**Pull request merged**